core/src/main/scala/kafka/server/DelayedProduce.scala

```diff
@@ -124,6 +124,19 @@ class DelayedProduce(delayMs: Long,
   }
 }
 
+class SafeDelayedProduce(delayMs: Long,
```

Comment:
Actually I wonder if we need an extension. Perhaps the original `DelayedProduce` could take an optional lock in the constructor which defaults to `this` or just `new Object`?